### DISCUSSION TITLE
the ordering of html tag

### DISCUSSION THREAD

**Discussion by Yichen-fqyd** [pymupdf/PyMuPDF#1174] [Looking for help] [answered]
Hi
When converting pdf to html, sometimes the tag is ordered as natural reading order, sometime the order is quite off, what are the possible reasons for this, and are there any parameters that can control the output html tag order?

Thank you

**Comment by JorjMcKie** [comment, marked as answer]
Output variances of "(X)HTML" and "XML" are direct wrappers of MuPDF functions. Any theoretically reachable option to influence results can be inspected by looking at the response of the CLI command `mutool convert`.
As you will see, for these output types there is nothing specific.

For all I have seen in MuPDF source code, there is no special effort yet to arrange HTML in some reading order - like anywhere else in the library. There are good reasons for this: HTML content does also have position information  - similar to PDF. So the appearance would not profit from any such effort.

**Reply by Yichen-fqyd** [reply]
After the detailed inspection, I think it might be related to the page rotation issue, would it possible to adjust the tag position and any function that could identify the rotation matrix?

Attached the screen shots for converted html.

<img width="1023" alt="Screen Shot 2021-08-11 at 10 43 43 AM" src="https://user-images.githubusercontent.com/37560330/129050446-eda4bc56-6cdc-46f4-b448-bea24881ec82.png">

[16-18_AIG_Loss_Runs.pdf](https://github.com/pymupdf/PyMuPDF/files/6969244/16-18_AIG_Loss_Runs.pdf)

Thank you so much

**Reply by JorjMcKie** [reply]
No, as I wrote in a previous post:
This set of extractions is beyond my control, they are direct wrappers of MuPDF functions.
All you can do is experiment a little by adjusting the page rotation yourself.

**Reply by Yichen-fqyd** [reply]
If I wanna rotate the page by myself, do I just do page.set_rotation()? But the how should I also rotate the text bbox as well. I think I met the similar issue with this link: https://github.com/pymupdf/PyMuPDF/discussions/1069

**Comment by JorjMcKie** [comment]
Insertion of anything into a page always requires **_unrotated_** coordinates - i.e. they must be provided as if the page were not rotated.
Same is true for extraction coordinates (text, drawings, ...).
The only thing to observe for text insertion however is that you must provide a rotation angle for the text. If specifying 0 (or nothing), text (-box) insertion will be horizontal WRT the rotated page. So for a 90° rotated page, text will be inserted /appear bottom to top in the unrotated page, if specifying no angle.
